### ISSUE TITLE
Performance degradation in recent releases

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#4245]
**Describe the bug**

Recent versions of pyright have become untenably slow. Using 1.1.278 as a baseline, the next release is more than 3x as slow, and the most recent release doesn't terminate within 6 hours.

**To Reproduce**

```sh
$ npm install -g pyright@1.1.278
$ time poetry run pyright queryai_ocsf

Found 933 source files
pyright 1.1.278
0 errors, 0 warnings, 0 informations
Completed in 6.027sec
poetry run pyright queryai_ocsf  11.54s user 0.69s system 169% cpu 7.214 total
```

```sh
» npm install -g pyright@1.1.279
» time poetry run pyright queryai_ocsf

Found 933 source files
pyright 1.1.279
0 errors, 0 warnings, 0 informations
Completed in 22.63sec
poetry run pyright queryai_ocsf  27.90s user 0.78s system 120% cpu 23.773 total
```

```sh
» npm install -g pyright@1.1.281
» time poetry run pyright queryai_ocsf

Found 933 source files
^C
poetry run pyright queryai_ocsf  133.17s user 0.96s system 100% cpu 2:13.97 total
```

The most recent version (1.1.281) ran for 6 hours in our GitHub CI and still didn't finish. Normally this takes < 90 seconds.

**Additional context**

Our project has a lot of source files, but they are all `@dataclass` declarations, code to instantiate these classes, and code to test roundtrip serialization -- there's no advanced type system features being used, no generics, etc.

**Comment by kyle-query** [comment]
There was a noticeable slowdown (~3X longer) in 1.1.279, same runtime in 1.1.280, then a severe performance issue in 1.1.281.

I've uploaded the output from 276 to 280 [here](https://gist.github.com/kyle-query/590e959d0828da88d316a4a576110fa0).

The project is not open source, but I'm checking if I can share the relevant code with you. In summary, it's simple dataclass definitions, code to create random instances of those classes, and code to translate between those dataclasses and Google's `protoc`-generated classes (these use the `protobuf` library at runtime).

In the last unaffected version (1.1.278), the most time was spent checking `codec/` files. In 1.1.279 and 1.1.280, the most time was spent checking files in `random/`. The timing stats for `Check` increased from 6s to 24s, but the other timing stats were the same.

I'm currently running 1.1.281 to collect debug information and it's very slowly working through files in `codec/*`. It doesn't seem to be a single file at issue.

Currently pyright is consuming one CPU core completely, and the memory usage seems stable at < 250MB after running for an hour. Here's a sample of the output from 1.1.281:

```
[FG]   analyzing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py ...
[FG]     parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py [fs read 0ms] (3ms)
[FG]     binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py (1ms)
[FG]     checking: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py ...
[FG]       es ...
[FG]          ...
[FG]           parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/classes/access_activity.py [fs read 0ms] (3ms)
[FG]           binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/classes/access_activity.py (1ms)
[FG]          (4ms) [f:0, t:1, p:1, i:1, b:1]
[FG]          ...
[FG]           parsing: /Users/kyle/.asdf/installs/nodejs/18.4.0/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/dataclasses.pyi [fs read 0ms] (2ms)
[FG]           binding: /Users/kyle/.asdf/installs/nodejs/18.4.0/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/dataclasses.pyi (0ms)
[FG]          (3ms) [f:0, t:1, p:1, i:0, b:0]
[FG]         Ve ...
[FG]            ...
[FG]             parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/helper.py [fs read 0ms] (0ms)
[FG]             binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/helper.py (0ms)
[FG]            (0ms)
[FG]         Ve ["Optional[JsonVal]" (queryai_ocsf.dataclass.classes.access_activity) [46:11]] (1ms)
[FG]       es [ImportFromAs '"AccessActivity"  [12:5]' '"dc_AccessActivity"  [12:23]' (queryai_ocsf.codec.classes.access_activity) [12:5]] (13ms) [f:0, t:2, p:2, i:1, b:1]
[FG]       es ...
[FG]          ...
[FG]           parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/google/classes/access_activity_pb2.pyi [fs read 1ms] (2ms)
[FG]           binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/google/classes/access_activity_pb2.pyi (0ms)
[FG]          (2ms) [f:1, t:0, p:1, i:0, b:0]
[FG]       es [ImportFromAs '"AccessActivity"  [15:5]' '"pb_AccessActivity"  [15:23]' (queryai_ocsf.codec.classes.access_activity) [15:5]] (3ms) [f:1, t:0, p:1, i:0, b:0]
[FG]       ke ["dst.data.CopyFrom(encode_pb_js <shortened> " (queryai_ocsf.codec.classes.access_activity) [29:9]] (6044ms)
[FG] Long operation: ke (6044ms)
[FG]       es [Assignment (queryai_ocsf.codec.classes.access_activity) [67:9]] (84338ms)
[FG] Long operation: es (84338ms)
[FG]     checking: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py (90425ms)
[FG] Long operation: checking: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py (90425ms)
[FG]   analyzing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py (90429ms)
[FG] Long operation: analyzing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/access_activity.py (90429ms)
[FG]   analyzing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/account_change.py ...
[FG]     parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/account_change.py [fs read 1ms] (6ms)
[FG]     binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/account_change.py (0ms)
[FG]     checking: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/codec/classes/account_change.py ...
[FG]       es ...
[FG]          ...
[FG]           parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/classes/account_change.py [fs read 1ms] (3ms)
[FG]           binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/dataclass/classes/account_change.py (0ms)
[FG]          (3ms) [f:1, t:1, p:1, i:0, b:0]
[FG]       es [ImportFromAs '"AccountChange"  [12:5]' '"dc_AccountChange"  [12:22]' (queryai_ocsf.codec.classes.account_change) [12:5]] (6ms) [f:1, t:1, p:1, i:0, b:0]
[FG]       es ...
[FG]          ...
[FG]           parsing: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/google/classes/account_change_pb2.pyi [fs read 0ms] (1ms)
[FG]           binding: /Users/kyle/wd/query/query-ocsf/queryai_ocsf/google/classes/account_change_pb2.pyi (1ms)
[FG]          (2ms) [f:0, t:0, p:1, i:0, b:1]
[FG]       es [ImportFromAs '"AccountChange"  [15:5]' '"pb_AccountChange"  [15:22]' (queryai_ocsf.codec.classes.account_change) [15:5]] (2ms) [f:0, t:0, p:1, i:0, b:1]
[FG]       ke ["dst.data.CopyFrom(encode_pb_js <shortened> " (queryai_ocsf.codec.classes.account_change) [29:9]] (6111ms)
[FG] Long operation: ke (6111ms)
[FG]       es [Assignment (queryai_ocsf.codec.classes.account_change) [67:9]] (84318ms)
[FG] Long operation: es (84318ms)
```

There's at least a few lines in the log like this. What are `ke`, `es`, `we`, etc? Only `es` seems to appear in the debug output in older versions of pyright, but the `ke` lines for each source file in `codec/*` look suspicious:

```
[FG]       ke ["dst.data.CopyFrom(encode_pb_js <shortened> " (queryai_ocsf.codec.classes.dev.command_activity) [33:9]] (6021ms)
```

I assume this means that particular statement took 6 seconds to type check. `CopyFrom` is implemented by the [Google protobuf runtime](https://googleapis.dev/python/protobuf/latest/google/protobuf/message.html#google.protobuf.message.Message.CopyFrom). Here `dst.data` has the type `_json_pb2.JsonVal` which is generated by `protoc` from this protobuf definition:

```proto
syntax = "proto2";

package queryai;

message JsonVal {
    oneof val {
        bool    boolean  = 1;
        int64   integer  = 2;
        double  float    = 3;
        string  string   = 4;
        JsonObj object   = 5;
        JsonArr array    = 6;
    }
}

message JsonObj {
    map<string, JsonVal> items = 1;
}

message JsonArr {
    repeated JsonVal items = 2;
}
```

Filling in the `<shortened>` part, the whole statement is `dst.data.CopyFrom(encode_pb_jsonval(src.data))` where `src.data` has the type `_JsonVal`. That type is declared as so:

```python
from typing import Mapping, Optional, Sequence, Union

_JsonArr = Sequence[Optional["_JsonVal"]]
_JsonObj = Mapping[str, Optional["_JsonVal"]]
_JsonVal = Union[bool, float, int, str, "_JsonArr", "_JsonObj"]
```

And `encode_pb_jsonval` has the type `_json_pb2.JsonVal -> _JsonVal`, it is simply translating between the two representations of JSON-like data.

**Comment by erictraut** [comment]
OK, thanks for the additional details. I suspect that it's related to the recursive type alias that you're using for `_JsonVal`. I will likely need additional context from you about the expression that is taking so much time to evaluate.  If you're able to produce a self-contained code sample that exhibits the problem, that would be very helpful.

If you're interested in the details... Pyright contains some recursion checks that prevent infinite recursion when it is evaluating types. In version 1.1.281, I increased the recursion limit to eliminate a bug that was caused when the limit was hit for a protocol class that involved a self reference. I suspect that you're hitting this recursion limit in your code. I'll need to determine how to repro this case and mitigate the recursion to avoid the long analysis time.

**Comment by kyle-query** [comment]
Ok, I obtained permission to share a large subset of the code with you. I'll setup a private repo with the relevant files for you Thursday or Friday.

I also updated the linked gist with the debug logs from [1.1.281](https://gist.githubusercontent.com/kyle-query/590e959d0828da88d316a4a576110fa0/raw/d8cc6bbf49ce3202a9ebc8ceaf4ad05a0ab3c65d/1.1.281.txt). It shows a strongly skewed distribution of time per file, with most files at < 150ms and the others above 80,000ms.

I won't have time until later this week to look more closely at the results, but a quick sample suggested you might be right -- the files that were checked quickly didn't make use of that recursive type and the problematic files did.

**Comment by erictraut** [comment]
Thanks @kyle-query. I've figured out the root cause of the problem.

The problem is rooted in the fact that you've declared two sets of recursive type aliases (both called `JsonArr`, `JsonObj`, and `JsonVal`) in two different files. Pyright has special logic to detect and shortcut recursion when determining type compatibility between two types that use the same recursive type alias. You  are using two different recursive type aliases, so these recursion checks don't apply, and it continues to evaluate the types until it hits an internal recursion limit. This can take a long time.

The simple workaround in your code is to avoid defining multiple copies of these recursive type aliases. The two files in question are `dataclass/json.py` and `random/helper.py`. I modified `random/helper.py` to import (and re-export) the type aliases from `dataclass/json.py` rather defining its own copy. Type checking for your entire source base (at least the 963 source files you shared with me privately) now completes in just 2.8sec on my machine.

I'll look for ways to detect and fix the condition you were hitting.

**Comment by erictraut** [comment]
I've added [code to detect this condition](https://github.com/microsoft/pyright/commit/3ba6f3beec6ecbced192d647056ed57eeab12311) and mitigate the performance impact. With my change, type analysis of your code base goes from "effectively infinite" to 5.3sec on my machine. If you remove the duplicate definitions of `JsonVal`, this further drops to 2.8sec.

This improvement will be in the next release of pyright.

Thanks for sharing your code with me. It would have been much more difficult and time-consuming to figure out the problem without this. You can rescind the permissions to your repo, as I don't need further access.

**Comment by kyle-query** [comment]
Thank you Eric for the super quick turn around time; and for providing a short-term workaround until next release!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.283, which I just published. It will also be included in a future version of pylance.
